cli/internal/run/global_hash.go

```diff
@@ -47,28 +47,40 @@ type oldGlobalHashable struct {
 }
 
 // calculateGlobalHashFromHashable returns a hash string from the globalHashable
-func calculateGlobalHashFromHashable(named GlobalHashable) (string, error) {
-	// When we aren't in infer mode, we can hash the whole object
-	if named.envMode != util.Infer {
-		return fs.HashObject(named)
-	}
+func calculateGlobalHashFromHashable(full GlobalHashable) (string, error) {
+	switch full.envMode {
+	case util.Infer:
+		if full.envVarPassthroughs != nil {
+			// In infer mode, if there is any passThru config (even if it is an empty array)
+			// we'll hash the whole object, so we can detect changes to that config
+			// Further, resolve the envMode to the concrete value.
+			full.envMode = util.Strict
+			return fs.HashObject(full)
+		}
 
-	// In infer mode, if there is any passThru config (even if it is an empty array)
-	// we'll hash the whole object, so we can detect changes to that config
-	if named.envVarPassthroughs != nil {
-		return fs.HashObject(named)
+		// If we're in infer mode, and there is no global pass through config,
+		// we use the old struct layout. this will be true for everyone not using the strict env
+		// feature, and we don't want to break their cache.
+		return fs.HashObject(oldGlobalHashable{
```

Comment:
I know I missed the boat here but is my assumption that `oldGlobalHashable` == 'non-strict mode' correct? If that is the case, then maybe renaming to reflect that would be an idea.

Reply:
There have been iterations on this where it used to be not "just env mode" ... we got here by path dependency and the external opinion is a good thought.